astropy/wcs/src/util.c

```diff
@@ -20,18 +20,28 @@ void set_invalid_to_nan(
   const int* s = stat;
   const int* s_end = stat + ncoord;
   double n;
+  int bit;
 
   #ifndef NAN
     #define INF (DBL_MAX+DBL_MAX)
     #define NAN (INF-INF)
   #endif
 
+  // Note that stat is a bit mask, so we need to mask only some of
+  // the coordinates depending on the bit mask values.
+
   n = NAN;
 
   for ( ; s != s_end; ++s) {
     if (*s) {
+      bit = 1;
       for (i = 0; i < nelem; ++i) {
-        *d++ = n;
+        if ((*s & bit) == bit) {
+          *d++ = n;
```

Comment:
Even if it may sound stupid: What does this line do? 😕

Reply:
Just so you know, I was wondering the same thing!

Reply:
I *think* it basically increments the pointer to the data, then dereferences it and assigns it to NaN all in one go? It was written by the great wizard @mdboom back in the days :)

Reply:
I think it's just doing:

```
d++;
*d = n
```

Reply:
The other way around, I think:
```
*d = n;
d++;
```
(otherwise it would be `*++d`)

Reply:
I.e., it sets the result to `nan` and then goes to the next one, while in the `else` clause one just goes to the next one without making a change. Personally, I think it would be way clearer to write `*d = n` here and then not have an else clause and either do `d++` below or in the `for` statement.  But out of scope here!

Reply:
Ah yeah that makes more sense :)

Reply:
Thank you for the clarification!

> Personally, I think it would be way clearer to write `*d = n` here and then not have an else clause and either do `d++` below or in the for statement. But out of scope here!

I don't think it would be out of scope we already touch these lines and the `else` wasn't present before. It's subjective but what you propose would make it much clearer to me.

Reply:
Ok I've done this now and it is indeed cleaner, thanks!